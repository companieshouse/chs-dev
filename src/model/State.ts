/**
 * Represents the state of the world for the environment
 */
export interface State {

    /**
     * list of module names of the modules the user has enabled
     */
    modules: string[];

    /**
     * list of the service names the user has enabled
     */
    services: string[];

    /**
     * list of the service names which are within development mode within the
     * environment
     */
    servicesWithLiveUpdate: string[];

    /**
     * list of service names which are removed from the deployed environment
     */
    excludedServices: string[];
}

export default State;
