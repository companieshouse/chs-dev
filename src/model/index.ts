export enum LogCoverage {
  WATCH,
  LOG,
}

export enum Prune {
  VOLUME,
  NETWORK,
  IMAGE,
  CONTAINER,
  BUILDER,
}
