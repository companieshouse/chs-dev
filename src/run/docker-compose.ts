import { execSync, spawn } from "child_process";
import { existsSync, mkdirSync, readFileSync } from "fs";
import { join } from "path";
import { LogHandler } from "./logs/logs-handler.js";
import DockerComposeWatchLogHandler from "./logs/DockerComposeWatchLogHandler.js";
import PatternMatchingConsoleLogHandler from "./logs/PatternMatchingConsoleLogHandler.js";

interface Logger {
    log: (msg: string) => void;
}

const CONTAINER_STOPPED_STATUS_PATTERN =
    /Container\s([\dA-Za-z-]*)\s*(Stopped|Removed)/;

const CONTAINER_STARTED_HEALTHY_STATUS_PATTERN =
    /Container\s([\dA-Za-z-]*)\s*(Started|Healthy|Stopped)/;

export class DockerCompose {

    readonly logFile: string;

    constructor (readonly path: string, readonly logger: Logger) {
        const logsDir = join(this.path, "local/.logs");
        if (!existsSync(logsDir)) {
            mkdirSync(
                logsDir,
                {
                    recursive: true
                }
            );
        }

        this.logFile = join(logsDir,
            `compose.out.${Math.floor(Date.now() / 1000)}.txt`);
    }

    down (signal?: AbortSignal): Promise<void> {
        return this.dockerComposeAction(["down", "--remove-orphans"],
            this.createStatusMatchLogHandler(CONTAINER_STOPPED_STATUS_PATTERN),
            signal
        );
    }

    getServiceStatuses (): Record<string, string> | undefined {
        if (!existsSync(join(this.path, "docker-compose.yaml"))) {
            return undefined;
        }

        // Call docker compose and output the services and their status as a CSV
        // each service on an individual line
        const dockerComposePsOutput = execSync(
            "docker compose ps -a --format '{{.Service}},{{.Status}}' 2>/dev/null || : \"\"",
            { cwd: this.path }
        ).toString("utf-8");

        // Parse the compose output and load into a Record i.e. service name to
        // status
        return dockerComposePsOutput.length > 0
            ? Object.fromEntries(dockerComposePsOutput.split("\n")
                .map(line => line.split(",")))
            : undefined;
    }

    up (signal?: AbortSignal): Promise<void> {
        return this.dockerComposeAction(["up", "-d", "--remove-orphans"],
            this.createStatusMatchLogHandler(CONTAINER_STARTED_HEALTHY_STATUS_PATTERN),
            signal
        );
    }

    watch (signal?: AbortSignal): Promise<void> {
        return this.dockerComposeAction([
            "watch"
        ],
        new DockerComposeWatchLogHandler(this.logFile, this.logger),
        signal
        );
    }

    private dockerComposeAction (
        composeArgs: string[],
        logHandler: LogHandler,
        signal?: AbortSignal
    ): Promise<void> {
        const dataHandler = (data: string) => logHandler.handle(data);

        return this.runDockerCompose(composeArgs, dataHandler, signal);
    }

    private createStatusMatchLogHandler (pattern: RegExp) {
        return new PatternMatchingConsoleLogHandler(
            pattern, this.logFile, this.logger
        );
    }

    private runDockerCompose (composeArgs: string[], listener: (chunk: any) => void, signal?: AbortSignal): Promise<void> {
        return new Promise((resolve, reject) => {
            // Spawn docker compose process

            const sshPrivateKey = this.sshPrivateKey();
            let spawnArgs: {
                cwd: string,
                signal?: AbortSignal,
                env?: Record<string, string>
            } = {
                cwd: this.path,
                signal
            };

            if (sshPrivateKey) {
                spawnArgs = {
                    ...spawnArgs,
                    env: {
                        ...(process.env),
                        SSH_PRIVATE_KEY: sshPrivateKey
                    }
                };
            }

            const dockerComposeProcess = spawn(
                "docker",
                [
                    "compose",
                    ...composeArgs
                ],
                spawnArgs
            );

            // Handle log statements to stdout and stderr
            dockerComposeProcess.stdout.on("data", listener);
            dockerComposeProcess.stderr.on("data", listener);

            // Handle exits and errors by resolving/rejecting the promise accordingly
            dockerComposeProcess.once("exit", (code: number, signal: string) => {
                if (code === 0 || code === 130) {
                    resolve();
                } else {
                    reject(new Error(`Docker compose up exited with code: ${code}`));
                }
            });

            dockerComposeProcess.once("error", (err: Error) => {
                reject(err);
            });
        });
    }

    private sshPrivateKey (): string | undefined {
        if (!("CHS_DEV_GITHUB_SSH_PRIVATE_KEY" in process.env)) {
            return undefined;
        }

        return readFileSync(process.env.CHS_DEV_GITHUB_SSH_PRIVATE_KEY as string).toString("utf8");
    }
}
