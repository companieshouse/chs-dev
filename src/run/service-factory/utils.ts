import NewServiceSpec from "../../model/NewServiceSpec.js";

/**
 * Apply the value to the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param selector Jq like selector to apply the value to
 * @param newServiceSpec The newServiceSpec to apply the value to
 * @param newValue The value to apply
 * @returns The newServiceSpec with the value applied
 */
export const applySelector = (newServiceSpec: Partial<NewServiceSpec>, selector: string, newValue: any) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return {
            ...newServiceSpec,
            [nestedKeys[0]]: applySelector(
                newServiceSpecSubAttribute,
                `.${nestedKeys.slice(1).join(".")}`,
                newValue
            )
        };
    } else {
        return {
            ...newServiceSpec,
            [nestedKeys[0]]: newValue
        };
    }
};

/**
 * Lookup a value in the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param newServiceSpec NewServiceSpec to lookup the value in
 * @param selector Jq like selector to lookup the value at
 * @returns The value at the selector
 */
export const get = (newServiceSpec: Partial<NewServiceSpec>, selector: string) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return get(newServiceSpecSubAttribute, `.${nestedKeys.slice(1).join(".")}`);
    } else {
        return newServiceSpec[nestedKeys[0]];
    }
};

const isSelectorValid = (selector: string) => {
    return /^\.([A-Za-z0-9_-]+\.?)+$/.test(selector);
};

const extractValidNestedKeys = (selector: string) => {
    if (!isSelectorValid(selector)) {
        throw new Error("Invalid selector");
    }

    return selector.split(".").slice(1);
};

const getNextLevelObject = (newServiceSpec: Partial<NewServiceSpec>, nestedKey: string) => {
    return newServiceSpec[nestedKey] || {};
};
