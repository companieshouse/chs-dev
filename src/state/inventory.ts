import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "fs";
import { basename, join } from "path";

import yaml from "yaml";
import glob from "glob";

import { createHash } from "crypto";
import { Service } from "../model/Service.js";
import { Module } from "../model/Module.js";
import { readServices } from "./service-reader.js";
import { DependencyNameResolver } from "./dependency-name-resolver.js";

interface InventoryCache {
  hash: string;
  services: Service[];
  modules: Module[]
}

export class Inventory {
    readonly inventoryCacheFile: string;
    private __serviceFiles: string[] | undefined;

    constructor (private path: string, cacheDir: string) {
        this.path = path;
        this.inventoryCacheFile = join(cacheDir, `${basename(path)}.inventory.yaml`);

        if (!existsSync(cacheDir)) {
            mkdirSync(cacheDir, {
                recursive: true
            });
        }
    }

    get modules (): Module[] {
        return this.getFromCache(inventoryCache => inventoryCache.modules);
    }

    get services (): Service[] {
        return this.getFromCache(inventoryCache => inventoryCache.services);
    }

    private getFromCache<T> (cacheSupplier: (inventoryCache: InventoryCache) => T): T {
        const inventoryCache = this.inventoryCache;

        if (typeof inventoryCache !== "undefined") {
            if (inventoryCache.hash === this.hashServiceFiles()) {
                return cacheSupplier(inventoryCache);
            }
        }

        this.updateCache();

        return this.getFromCache(cacheSupplier);
    }

    private updateCache () {
        const inventory = {
            hash: this.hashServiceFiles(),
            services: this.loadServices(),
            modules: this.loadModules()
        };

        writeFileSync(this.inventoryCacheFile, yaml.stringify(inventory));
    }

    private loadModules (): Module[] {
        return readdirSync(join(this.path, "services", "modules"), { withFileTypes: true })
            .filter(item => item.isDirectory())
            .map(item => {
                return {
                    name: item.name
                };
            });
    }

    private loadServices (): Service[] {
        const partialServices: Partial<Service>[] = this.serviceFiles.flatMap(readServices);

        const dependencyNameResolver = new DependencyNameResolver(partialServices);

        return partialServices.map(service => ({
            ...service,
            dependsOn: dependencyNameResolver.fullDependencyListIncludingTransitive(service.dependsOn as string[])
        } as Service));
    }

    private hashServiceFiles () {
        const sha256Hash = createHash("sha256");

        this.serviceFiles
            .map(serviceFile => readFileSync(serviceFile))
            .forEach((serviceFile:any) => {
                sha256Hash.update(serviceFile);
            });

        return sha256Hash.digest("hex");
    }

    private get inventoryCache (): InventoryCache | undefined {
        if (existsSync(this.inventoryCacheFile)) {
            return yaml.parse(readFileSync(this.inventoryCacheFile).toString("utf-8"));
        }

        return undefined;
    }

    private get serviceFiles (): string[] {
        if (typeof this.__serviceFiles === "undefined") {
            this.__serviceFiles = glob.sync(`${this.path}/services/**/*.docker-compose.yaml`);
        }
        return this.__serviceFiles;
    }
}
