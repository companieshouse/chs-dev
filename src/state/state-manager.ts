import { existsSync, readFileSync, writeFileSync } from "fs";
import { join } from "path";
import yaml from "yaml";
import { State } from "../model/State.js";

const fileName = ".chs-dev.yaml";

export class StateManager {
    constructor (private path: string) {
        this.path = path;
    }

    get snapshot (): State {
        const path = join(this.path, fileName);
        if (existsSync(path)) {
            return yaml.parse(readFileSync(path).toString());
        }
        return {
            modules: [],
            services: [],
            servicesWithLiveUpdate: [],
            excludedServices: []
        };
    }

    includeModule (moduleName: string): void {
        const snapshot: State = this.snapshot;

        if (!snapshot.modules.includes(moduleName)) {
            this.dumpState({ ...snapshot, modules: [...snapshot.modules, moduleName] });
        }
    }

    excludeModule (moduleName: string): void {
        const snapshot: State = this.snapshot;

        if (snapshot.modules.includes(moduleName)) {
            this.dumpState({ ...snapshot, modules: snapshot.modules.filter(candidate => candidate !== moduleName) });
        }
    }

    includeService (serviceName: string): void {
        const snapshot: State = this.snapshot;

        if (!snapshot.services.includes(serviceName)) {
            this.dumpState({ ...snapshot, services: [...snapshot.services, serviceName] });
        }
    }

    excludeService (serviceName: string): void {
        const snapshot: State = this.snapshot;

        if (snapshot.services.includes(serviceName)) {
            this.dumpState({ ...snapshot, services: snapshot.services.filter(candidate => candidate !== serviceName) });
        }
    }

    async addExclusionForService (service: string): Promise<void> {
        let snapshot: State = this.snapshot;

        if (!snapshot.excludedServices) {
            this.dumpState({ ...snapshot, excludedServices: [] });
            snapshot = this.snapshot;
        }

        if (!snapshot.excludedServices.includes(service)) {
            this.dumpState({ ...snapshot, excludedServices: [...snapshot.excludedServices, service] });
        }
    }

    removeExclusionForService (service: string): void {
        let snapshot: State = this.snapshot;

        if (!snapshot.excludedServices) {
            this.dumpState({ ...snapshot, excludedServices: [] });
            snapshot = this.snapshot;
        }

        if (snapshot.excludedServices.includes(service)) {
            this.dumpState({ ...snapshot, excludedServices: snapshot.excludedServices.filter(candidate => candidate !== service) });
        }
    }

    includeServiceInLiveUpdate (serviceName: string): void {
        const snapshot: State = this.snapshot;

        if (!snapshot.servicesWithLiveUpdate.includes(serviceName)) {
            this.dumpState({ ...snapshot, servicesWithLiveUpdate: [...snapshot.servicesWithLiveUpdate, serviceName] });
        }
    }

    excludeServiceFromLiveUpdate (serviceName: string): void {
        const snapshot: State = this.snapshot;

        if (snapshot.servicesWithLiveUpdate.includes(serviceName)) {
            this.dumpState({ ...snapshot, servicesWithLiveUpdate: snapshot.servicesWithLiveUpdate.filter(candidate => candidate !== serviceName) });
        }
    }

    private dumpState (state: State): void {
        const lines = [
            "# DO NOT MODIFY MANUALLY",
            yaml.stringify(state)
        ];
        writeFileSync(join(this.path, fileName), lines.join("\n\n"));
    }
}
