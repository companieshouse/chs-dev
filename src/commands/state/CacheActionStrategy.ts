import yaml from "yaml";

import { getGeneratedDockerComposeFile } from "../../helpers/docker-compose-file.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import { writeContentToFile } from "../../helpers/file-utils.js";
import { handlePrompt, validateCacheNameExists } from "./state-cache-utils.js";
import { StateManager } from "../../state/state-manager.js";

const EXPORT_STATE_DIR = ".exported_state_cache";

export interface AvailableCacheContext {
    logger: (message: string) => void;
  }

export interface WipeCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface RemoveCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface ExportCacheContext {
    projectPath: string;
    logger: (message: string) => void;
}

export interface AddCacheContext {
    stateManager: StateManager;
    stateCacheFile: string;
    projectPath: string;
    logger: (message: string) => void;
}
interface CacheActionStrategy<C> {
    execute(cache: C, cacheData: Record<string, any>, cacheName?: string): Promise<void> | void;
}

const handleExportCache = (context: ExportCacheContext, cacheData: Record<string, any>, cacheName: string): void => {
    const exportDir = join(context.projectPath, EXPORT_STATE_DIR);
    if (!existsSync(exportDir)) {
        mkdirSync(exportDir, { recursive: true });
    }

    const exportedFilenamePath = join(exportDir, `${cacheName}.yaml`);

    writeContentToFile(cacheData[cacheName], exportedFilenamePath);
    context.logger(`Exported cache ${cacheName} destination: '${exportedFilenamePath}'`);
};

export class AvailableCacheStrategy implements CacheActionStrategy<AvailableCacheContext> {
    execute (context: AvailableCacheContext, cacheData: Record<string, any>): void {
        const cacheNames = Object.keys(cacheData);
        if (cacheNames.length > 0) {
            context.logger("Available caches:");
            cacheNames.forEach((name: string) => context.logger(`- ${name}`));
        } else {
            context.logger("No cache available.");
        }
    }
}

export class WipeCacheStrategy implements CacheActionStrategy<WipeCacheContext> {
    async execute (context: WipeCacheContext, _cacheData: Record<string, any>): Promise<void> {
        const confirmed = await handlePrompt("wipe");
        if (confirmed) {
            writeContentToFile({}, context.stateCacheFile);
            context.logger("Wiped all caches");
        }
    }
}

export class RemoveCacheStrategy implements CacheActionStrategy<RemoveCacheContext> {
    async execute (context: RemoveCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);

        const confirmed = await handlePrompt("remove", _cacheName);
        if (confirmed) {
            delete _cacheData[_cacheName];
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Removed cache '${_cacheName}'`);

        }
    }
}

export class ExportCacheStrategy implements CacheActionStrategy<ExportCacheContext> {
    async execute (context: ExportCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);
        handleExportCache(context, _cacheData, _cacheName);

    }
}

export class AddCacheStrategy implements CacheActionStrategy<AddCacheContext> {
    async execute (context: AddCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        const confirmed = await handlePrompt("add", _cacheName);
        if (confirmed) {
            const stateSnapshot = context.stateManager.snapshot;
            const dockerComposeSnapshot = getGeneratedDockerComposeFile(context.projectPath);
            _cacheData[_cacheName] = {
                state: {
                    hash: hashAlgorithm(yaml.stringify(stateSnapshot)),
                    snapshot: stateSnapshot
                },
                dockerCompose: {
                    hash: hashAlgorithm(yaml.stringify(dockerComposeSnapshot)),
                    snapshot: dockerComposeSnapshot
                }
            };
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Saved cache as '${_cacheName}'`);
        }
    }
}
