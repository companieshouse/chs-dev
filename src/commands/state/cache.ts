import { Args, Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Cache extends Command {

    static description = "Cache the state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { args, flags } = await this.parse(Cache);
        const timestamp: string = new Date(Date.now()).toISOString();
        const cacheName: string = args.name || timestamp;

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`save chs-dev state?`);
        };

        if (await handlePrompt()) {
            //
        }
    }

}
