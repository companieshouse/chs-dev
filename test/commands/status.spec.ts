import { jest, expect } from "@jest/globals";
import Status from "../../src/commands/status";
import { Config } from "@oclif/core";
import { State } from "../../src/model/State";
import { services, modules } from "../utils/data";
import { Inventory } from "../../src/state/inventory";
import { StateManager } from "../../src/state/state-manager";
import { DockerCompose } from "../../src/run/docker-compose";
import loadConfigMock from "../../src/helpers/config-loader.js";

const snapshot: State = {
    modules: [
        "module-one"
    ],
    services: [
        "service-one",
        "service-two",
        "service-five",
        "service-six",
        "service-seven",
        "service-eight",
        "service-three"
    ],
    servicesWithLiveUpdate: [
        "service-two"
    ],
    excludedServices: [
        "service-three"
    ]
};

const getServiceStatusesMock = jest.fn();

jest.mock("../../src/state/inventory");

jest.mock("../../src/state/state-manager");

jest.mock("../../src/run/docker-compose");

jest.mock("../../src/helpers/config-loader");

describe("Status command", () => {
    let status: Status;
    let testConfig: Config;
    let logMock;
    let logJsonMock;
    let runHookMock;
    let parseMock;

    beforeEach(() => {
        jest.resetAllMocks();

        // @ts-expect-error
        loadConfigMock.mockReturnValue({
            projectPath: "/users/user/docker-chs/"
        });

        runHookMock = jest.fn();

        // @ts-expect-error
        testConfig = { root: "./", configDir: "./config", cacheDir: "./cache", runHook: runHookMock };

        // @ts-expect-error
        Inventory.mockReturnValue({
            services,
            modules
        });

        // @ts-expect-error
        StateManager.mockReturnValue({
            snapshot
        });

        // @ts-expect-error
        DockerCompose.mockReturnValue({
            getServiceStatuses: getServiceStatusesMock
        });

        status = new Status([], testConfig);
        logMock = jest.spyOn(status, "log");
        // @ts-expect-error
        logJsonMock = jest.spyOn(status, "logJson");

        // @ts-expect-error
        parseMock = jest.spyOn(status, "parse");

        parseMock.mockResolvedValue({
            flags: {
                json: false
            }
        });
    });

    it("should log without docker compose statuses", async () => {
        await status.run();

        expect(logMock.mock.calls).toMatchSnapshot();
        expect(logJsonMock).not.toHaveBeenCalled();
    });

    it("should log with their docker compose statuses correctly", async () => {
        getServiceStatusesMock.mockReturnValue({
            "service-one": "Up 3 seconds (unhealthy)",
            "service-two": "Up 2 minutes (healthy)",
            "service-four": "Up 1 minute (health: starting)",
            "service-five": "Exited (0)",
            "service-six": "Exited (137)",
            "service-seven": "Up 33 minutes",
            "service-three": "Up About an hour (healthy)",
            "service-eight": "Up About an hour"
        });

        await status.run();

        expect(logMock.mock.calls).toMatchSnapshot();
    });

    it("should log json correctly", async () => {
        parseMock.mockResolvedValue({
            flags: {
                json: true
            }
        });
        getServiceStatusesMock.mockReturnValue({
            "service-one": "running",
            "service-two": "running",
            "service-five": "stopped",
            "service-four": "stopped"
        });

        await status.run();

        expect(logJsonMock).toHaveBeenCalledTimes(1);
        expect(logJsonMock.mock.calls[0]).toMatchSnapshot();

        expect(logMock).not.toHaveBeenCalled();
    });

    describe("transient services in development mode", () => {
        beforeEach(() => {
            jest.resetAllMocks();

            // @ts-expect-error
            loadConfigMock.mockReturnValue({
                projectPath: "/users/user/docker-chs/"
            });

            // @ts-expect-error
            StateManager.mockReturnValue({
                snapshot: {
                    services: [
                        "service-one",
                        "service-seven"
                    ],
                    modules: ["module-three"],
                    servicesWithLiveUpdate: [
                        "service-six"
                    ],
                    excludedServices: [
                        "service-three"
                    ]
                }
            });

            // @ts-expect-error
            Inventory.mockReturnValue({
                services,
                modules
            });

            getServiceStatusesMock.mockReturnValue({
                "service-one": "running",
                "service-two": "running",
                "service-five": "stopped",
                "service-four": "stopped"
            });

            status = new Status([], testConfig);

            logMock = jest.spyOn(status, "log");

            // @ts-expect-error
            logJsonMock = jest.spyOn(status, "logJson");

            // @ts-expect-error
            parseMock = jest.spyOn(status, "parse");
        });

        it("should return correct log without six in development mode", async () => {
            parseMock.mockResolvedValue({
                flags: {
                    json: false
                }
            });
            await status.run();

            expect(logMock.mock.calls).toMatchSnapshot();
        });

        it("should show not show a service in live update if is not enabled", async () => {

            parseMock.mockResolvedValue({
                flags: {
                    json: true
                }
            });
            await status.run();

            expect(logJsonMock).toHaveBeenCalledTimes(1);
            expect(logJsonMock.mock.calls[0]).toMatchSnapshot();
        });
    });
});
