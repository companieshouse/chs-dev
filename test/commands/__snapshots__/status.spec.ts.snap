// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status command should log json correctly 1`] = `
[
  {
    "modules": [
      "module-one",
    ],
    "services": [
      {
        "composeStatus": "stopped",
        "liveUpdate": false,
        "name": "service-five",
        "transient": true,
      },
      {
        "composeStatus": "stopped",
        "liveUpdate": false,
        "name": "service-four",
        "transient": true,
      },
      {
        "composeStatus": "running",
        "liveUpdate": false,
        "name": "service-one",
        "transient": false,
      },
      {
        "composeStatus": "running",
        "liveUpdate": true,
        "name": "service-two",
        "transient": false,
      },
    ],
  },
]
`;

exports[`Status command should log with their docker compose statuses 1`] = `
[
  [
    "Manually activated modules:",
  ],
  [
    " - module-one",
  ],
  [
    "
Manually activated services:",
  ],
  [
    " - service-one (running)",
  ],
  [
    " - service-two (running)",
  ],
  [
    "
Automatically activated services:",
  ],
  [
    " - service-five (stopped) ",
  ],
  [
    " - service-four (stopped) ",
  ],
  [
    " - service-one (running) ",
  ],
  [
    " - service-two (running) [LIVE UPDATE]",
  ],
  [
    "
Manually deactivated services:",
  ],
  [
    " - service-three",
  ],
]
`;

exports[`Status command should log without docker compose statuses 1`] = `
[
  [
    "Manually activated modules:",
  ],
  [
    " - module-one",
  ],
  [
    "
Manually activated services:",
  ],
  [
    " - service-one ",
  ],
  [
    " - service-two ",
  ],
  [
    "
Automatically activated services:",
  ],
  [
    " - service-five  ",
  ],
  [
    " - service-four  ",
  ],
  [
    " - service-one  ",
  ],
  [
    " - service-two  [LIVE UPDATE]",
  ],
  [
    "
Manually deactivated services:",
  ],
  [
    " - service-three",
  ],
]
`;
